### PR TITLE
Add AsyncRequest

### PR DESCRIPTION
**Pull request by pierre-emmanuelJ**
- [ ] example in `doc.go`

### CODE REVIEW

**Comment by pierre-emmanuelJ** [comment]
I'll write some test asap

File: request.go [hunk 1]
```diff
@@ -203,6 +203,131 @@ func (exo *Client) RequestWithContext(ctx context.Context, request Command) (int
 	}
 }
 
+// AsyncRequest performs the given command
+func (exo *Client) AsyncRequest(request Command, callback WaitAsyncJobResultFunc) {
+	ctx, cancel := context.WithTimeout(context.Background(), exo.Timeout)
+	defer cancel()
+
+	switch request.(type) {
+	case syncCommand:
```
Comment (greut) [inline request.go]: `AsyncRequest` shouldn't accept that. It's maybe time to make `syncCommand` and `asyncCommand` public again.

File: client_type.go [hunk 2]
```diff
@@ -55,3 +55,6 @@ type RetryStrategyFunc func(int64) time.Duration
 
 // IterateItemFunc represents the callback to iterate a list of results, if false stops
 type IterateItemFunc func(interface{}, error) bool
+
+// WaitAsyncJobResultFunc represents the callback to wait a results of an async request, if false stops
+type WaitAsyncJobResultFunc func(interface{}, interface{}, error) bool
```
Comment (greut) [inline client_type.go]: `func(JobStatus, interface{}, error)` ?

Comment (greut) [reply client_type.go]: or we find a way to put everything into `JobResult` and postpone the JSON unmarshalling

```go
jobResult.JobStatus
jobResult.Result().(*egoscale.VirtualMachine)
```

File: request.go (hunk 1 above)
Comment (pierre-emmanuelJ) [reply request.go]: 👍

File: client_type.go (hunk 2 above)
Comment (pierre-emmanuelJ) [reply client_type.go]: Yes it's really better like that !

File: request.go (hunk 1 above)
Comment (greut) [reply request.go]: or only _async_

File: client_type.go (hunk 2 above)
Comment (greut) [reply client_type.go]:
```golang
vm := &VirtualMachine{}
if err := asyncJobResult.Result(vm); err {
    ...
}
```

**Comment by greut** [comment]
```golang
req := &DeployVirtualMachine{...}
vm := &VirtualMachine{}
client.AsyncRequest(req, func(result AsyncJobResult, e error) bool {
    // Failure goes here
    if e != nil {
        panic(e)
    }  
    if (result.JobStatus == Pending) {
        log.Printf("waiting... on %s", result.JobID)
        return true
    }
    if err := result.Result(vm); err != nil
        panic(e)
    }
    
    return false
})
```

File: client_type.go [hunk 3]
```diff
@@ -55,3 +55,6 @@ type RetryStrategyFunc func(int64) time.Duration
 
 // IterateItemFunc represents the callback to iterate a list of results, if false stops
 type IterateItemFunc func(interface{}, error) bool
+
+// WaitAsyncJobResultFunc represents the callback to wait a results of an async request, if false stops
+type WaitAsyncJobResultFunc func(interface{}, error) bool
```
Comment (greut) [inline client_type.go]: `(AsyncJobResult, error)`

File: request.go [hunk 4]
```diff
@@ -196,13 +196,107 @@ func (exo *Client) RequestWithContext(ctx context.Context, request Command) (int
...
+	defer cancel()
+	exo.makeAsyncRequest(ctx, request, callback)
+}
+
+// AsyncRequestWithContext preforms a request with a context
+func (exo *Client) AsyncRequestWithContext(ctx context.Context, request AsyncCommand, callback WaitAsyncJobResultFunc) {
+	exo.makeAsyncRequest(ctx, request, callback)
+}
+
+func (exo *Client) makeAsyncRequest(ctx context.Context, request AsyncCommand, callback WaitAsyncJobResultFunc) {
```
Comment (greut) [inline request.go]: you may remove the lines 215-218

File: async_jobs_type.go [hunk 5]
```diff
@@ -18,7 +18,7 @@ type AsyncJobResult struct {
 	JobStatus       JobStatusType    `json:"jobstatus"`
 	UserID          string           `json:"userid"`
 	JobID           string           `json:"jobid"`
-	JobResponse     interface{}      `json:"jobresponse,imotempty"`
+	response        interface{}
```
Comment (greut) [inline async_jobs_type.go]: I don't think it's needed after all :-)

File: request.go [hunk 6]
```diff
@@ -271,23 +263,13 @@ func (exo *Client) makeAsyncRequest(ctx context.Context, request AsyncCommand, c
...
 		} else if res.JobStatus == Failure {
-			r := new(ErrorResponse)
-			if e := json.Unmarshal(*res.JobResult, r); e != nil {
-				if !callback(nil, e) {
-					return
-				}
-			}
-			if !callback(res, r) {
+			if !callback(res, nil) {
 				return
```
Comment (greut) [inline request.go]:
```go
if success {
    callback(res, nil)
} else if failure {
    callback(nil, res.Error())
}
```

**Comment by greut** [comment]
This PR is a superb addition/refactoring!

File: request_test.go [hunk 7]
```diff
@@ -198,7 +198,65 @@ func TestBooleanRequestTimeout(t *testing.T) {
...
+			if r := j.Response(resp); r != nil {
+				t.Error(r)
+			}
+			return false
+		}
+		return true
+	})
+
+	if resp.VirtualMachine.ServiceOfferingID != "71004023-bb72-4a97-b1e9-bc66dfce9470" {
+		t.Errorf("Expected serviofferingID %q, got %q", "71004023-bb72-4a97-b1e9-bc66dfce9470", resp.VirtualMachine.ServiceOfferingID)
```
Comment (greut) [inline request_test.go]: typo :warning:

File: client_test.go [hunk 8]
```diff
@@ -62,7 +62,7 @@ func TestClientAsyncDelete(t *testing.T) {
 	"jobresult": {
 		"success": false,
 		"displaytext": "herp derp",
-	},
+    },
```
Comment (greut) [inline client_test.go]: nope

File: async_jobs.go [hunk 9]
```diff
@@ -24,3 +26,24 @@ func (*ListAsyncJobs) name() string {
...
+	return nil
+}
+
+func (a *AsyncJobResult) Error() error {
+	r := new(ErrorResponse)
+	if e := json.Unmarshal(*a.JobResult, r); e != nil {
+		return e
+	}
+	return r
+}
```
Comment (greut) [inline async_jobs.go]: that one is never called in the tests, it's seems to be an easy reach

File: request_test.go [hunk 10]
```diff
@@ -256,6 +256,104 @@ func TestAsyncRequestWithoutContext(t *testing.T) {
...
+					"jobresult": {},
+					"jobstatus": 0
+				}
+				}
+				`}, response{200, `
+					{"queryasyncjobresultresponse": {
+						"jobid": "1",
+						"jobresult": [],
+						"jobstatus": 1
+				}}`})
```
Comment (greut) [inline request_test.go]: do you mind indenting all the responses at the same level?

File: request_test.go [hunk 11]
```diff
@@ -256,6 +256,104 @@ func TestAsyncRequestWithoutContext(t *testing.T) {
...
+			t.Errorf("Expected an error, got <nil>")
+		}
+		return true
+	})
+
+	cs.AsyncRequest(req, func(j *AsyncJobResult, err error) bool {
+		if err != nil {
+			return false
+		}
+		return true
```
Comment (greut) [inline request_test.go]: `return err == nil`

File: request_test.go [hunk 12]
```diff
@@ -256,6 +256,104 @@ func TestAsyncRequestWithoutContext(t *testing.T) {
...
+	})
+
+	cs.AsyncRequest(req, func(j *AsyncJobResult, err error) bool {
+		if err != nil {
+			t.Error(err)
```
Comment (greut) [inline request_test.go]: `Fatal`

Comment (pierre-emmanuelJ) [reply request_test.go]: Put Fatal instead of t.Error() ?

**Comment by pierre-emmanuelJ** [comment]
Yes sorry I forget go doc

**Pull request merged**